File: crates/dw-ado/src/urls.rs
use crate::{AzureDevOpsOptions, DEFAULT_API_VERSION};

pub fn expanded_work_item_url(options: &AzureDevOpsOptions, work_item_id: &str) -> String {
    format!(
        "{}/{}/_apis/wit/workitems/{}?$expand=all&api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        work_item_id,
        api_version(options)
    )
}

pub fn work_item_comments_url(
    options: &AzureDevOpsOptions,
    work_item_id: &str,
    top: u32,
) -> String {
    format!(
        "{}/{}/_apis/wit/workItems/{}/comments?$top={}&api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        work_item_id,
        top,
        api_version(options)
    )
}

pub fn work_item_url(options: &AzureDevOpsOptions, work_item_id: &str) -> String {
    format!(
        "{}/{}/_apis/wit/workitems/{}?api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        work_item_id,
        api_version(options)
    )
}

pub fn work_items_batch_url(options: &AzureDevOpsOptions) -> String {
    format!(
        "{}/{}/_apis/wit/workitemsbatch?api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        api_version(options)
    )
}

pub fn work_item_api_url(options: &AzureDevOpsOptions, work_item_id: &str) -> String {
    format!(
        "{}/{}/_apis/wit/workItems/{}",
        options.organization.trim_end_matches('/'),
        options.project,
        work_item_id
    )
}

pub fn work_item_web_url(options: &AzureDevOpsOptions, work_item_id: &str) -> String {
    format!(
        "{}/{}/_workitems/edit/{}",
        options.organization.trim_end_matches('/'),
        encode_component(&options.project),
        work_item_id
    )
}

pub fn create_work_item_url(options: &AzureDevOpsOptions, work_item_type: &str) -> String {
    format!(
        "{}/{}/_apis/wit/workitems/${}?api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        encode_component(work_item_type),
        api_version(options)
    )
}

pub fn pull_requests_url(options: &AzureDevOpsOptions, repository: &str) -> String {
    format!(
        "{}/{}/_apis/git/repositories/{}/pullrequests?api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        encode_component(repository),
        api_version(options)
    )
}

pub fn pull_request_web_url(
    options: &AzureDevOpsOptions,
    repository: &str,
    pull_request_id: i64,
) -> String {
    format!(
        "{}/{}/_git/{}/pullrequest/{}",
        options.organization.trim_end_matches('/'),
        encode_component(&options.project),
        encode_component(repository),
        pull_request_id
    )
}

pub fn active_pull_requests_url(
    options: &AzureDevOpsOptions,
    repository: &str,
    source_ref: &str,
) -> String {
    format!(
        "{}/{}/_apis/git/repositories/{}/pullrequests?searchCriteria.status=active&searchCriteria.sourceRefName={}&api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        encode_component(repository),
        encode_component(source_ref),
        api_version(options)
    )
}

pub fn active_pull_requests_for_repository_url(
    options: &AzureDevOpsOptions,
    repository: &str,
) -> String {
    format!(
        "{}/{}/_apis/git/repositories/{}/pullrequests?searchCriteria.status=active&api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        encode_component(repository),
        api_version(options)
    )
}

pub fn pull_request_work_items_url(
    options: &AzureDevOpsOptions,
    repository: &str,
    pull_request_id: i64,
) -> String {
    format!(
        "{}/{}/_apis/git/repositories/{}/pullRequests/{}/workitems?api-version={}",
        options.organization.trim_end_matches('/'),
        options.project,
        encode_component(repository),
        pull_request_id,
        api_version(options)
    )
}

pub(crate) fn organization_name(value: &str) -> String {
    let trimmed = value.trim().trim_end_matches('/');
    trimmed
        .rsplit('/')
        .next()
        .filter(|segment| !segment.trim().is_empty())
        .unwrap_or(trimmed)
        .to_owned()
}

fn api_version(options: &AzureDevOpsOptions) -> &str {
    if options.api_version.trim().is_empty() {
        DEFAULT_API_VERSION
    } else {
        &options.api_version
    }
}

fn encode_component(value: &str) -> String {
    value.replace(' ', "%20").replace('/', "%2F")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pull_request_web_url_targets_azure_devops_page_not_api_json() {
        let options = AzureDevOpsOptions {
            organization: "https://dev.azure.com/acme/".into(),
            project: "Hommage Agence".into(),
            api_version: "7.1".into(),
        };

        let url = pull_request_web_url(&options, "front app", 55264);

        assert_eq!(
            url,
            "https://dev.azure.com/acme/Hommage%20Agence/_git/front%20app/pullrequest/55264"
        );
        assert!(!url.contains("_apis"));
    }
}

File: internal/work/ado/urls.go
package ado

import (
	"net/url"
	"strconv"
	"strings"
)

func apiVersion(options Options) string {
	if strings.TrimSpace(options.APIVersion) == "" {
		return DefaultAPIVersion
	}
	return options.APIVersion
}

func adoURL(options Options, segments []string, query adoQuery) string {
	value, err := url.Parse(strings.TrimRight(options.Organization, "/"))
	if err != nil {
		value = &url.URL{Path: strings.TrimRight(options.Organization, "/")}
	}
	path := strings.TrimRight(value.Path, "/")
	escapedPath := strings.TrimRight(value.EscapedPath(), "/")
	for _, segment := range segments {
		path += "/" + segment
		escapedPath += "/" + url.PathEscape(segment)
	}
	value.Path = path
	value.RawPath = escapedPath
	value.RawQuery = query.encode()
	value.Fragment = ""
	return value.String()
}

type adoQuery []string

func (values adoQuery) encode() string {
	var result strings.Builder
	for index := 0; index < len(values); index += 2 {
		if index != 0 {
			result.WriteByte('&')
		}
		result.WriteString(strings.ReplaceAll(url.QueryEscape(values[index]), "%24", "$"))
		result.WriteByte('=')
		result.WriteString(url.QueryEscape(values[index+1]))
	}
	return result.String()
}

func query(values ...string) adoQuery {
	return values
}

func ExpandedWorkItemURL(options Options, id string) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workitems", id}, query("$expand", "all", "api-version", apiVersion(options)))
}

func WorkItemCommentsURL(options Options, id string, top int) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workItems", id, "comments"}, query("$top", strconv.Itoa(top), "api-version", CommentsAPIVersion))
}

func WorkItemURL(options Options, id string) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workitems", id}, query("api-version", apiVersion(options)))
}

func WorkItemsBatchURL(options Options) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workitemsbatch"}, query("api-version", apiVersion(options)))
}

func WIQLURL(options Options, top int) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "wiql"}, query("$top", strconv.Itoa(top), "api-version", apiVersion(options)))
}

func WorkItemAPIURL(options Options, id string) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workItems", id}, nil)
}

func WorkItemWebURL(options Options, id string) string {
	return adoURL(options, []string{options.Project, "_workitems", "edit", id}, nil)
}

func CreateWorkItemURL(options Options, workItemType string) string {
	return adoURL(options, []string{options.Project, "_apis", "wit", "workitems", "$" + workItemType}, query("api-version", apiVersion(options)))
}

func PullRequestsURL(options Options, repository string) string {
	return adoURL(options, []string{options.Project, "_apis", "git", "repositories", repository, "pullrequests"}, query("api-version", apiVersion(options)))
}

func PullRequestWebURL(options Options, repository string, id int64) string {
	return adoURL(options, []string{options.Project, "_git", repository, "pullrequest", strconv.FormatInt(id, 10)}, nil)
}

func PullRequestURL(options Options, repository string, id int64) string {
	return adoURL(options, []string{options.Project, "_apis", "git", "repositories", repository, "pullRequests", strconv.FormatInt(id, 10)}, query("api-version", apiVersion(options)))
}

func ActivePullRequestsURL(options Options, repository, sourceRef string) string {
	return adoURL(options, []string{options.Project, "_apis", "git", "repositories", repository, "pullrequests"}, query("searchCriteria.status", "active", "searchCriteria.sourceRefName", sourceRef, "api-version", apiVersion(options)))
}

func ActivePullRequestsForRepositoryURL(options Options, repository string, skip, top int) string {
	return adoURL(options, []string{options.Project, "_apis", "git", "repositories", repository, "pullrequests"}, query("searchCriteria.status", "active", "$skip", strconv.Itoa(skip), "$top", strconv.Itoa(top), "api-version", apiVersion(options)))
}

func PullRequestWorkItemsURL(options Options, repository string, id int64) string {
	return adoURL(options, []string{options.Project, "_apis", "git", "repositories", repository, "pullRequests", strconv.FormatInt(id, 10), "workitems"}, query("api-version", apiVersion(options)))
}

func ConnectionDataURL(options Options) string {
	return adoURL(options, []string{"_apis", "connectionData"}, query("connectOptions", "1", "lastChangeId", "-1", "lastChangeId64", "-1"))
}

func OrganizationName(value string) string {
	trimmed := strings.TrimRight(strings.TrimSpace(value), "/")
	parsed, err := url.Parse(trimmed)
	if err == nil {
		path := strings.Trim(parsed.Path, "/")
		if path != "" {
			parts := strings.Split(path, "/")
			return parts[len(parts)-1]
		}
		if parsed.Host != "" {
			return parsed.Host
		}
	}
	return trimmed
}
